File: C++/98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {        
public:
    bool isValidBST(TreeNode* root) {
        if (root == NULL)
            return true;        
        return isValidBST(root, INT_MIN, INT_MAX, false, false);             
    }
    
private:
    bool isValidBST(TreeNode* root, int leftMin, int rightMax, bool leftFlag, bool rightFlag){       
        if (root->left == NULL && root->right == NULL)
            return true;
        if (root->left == NULL)                    
            return root->right->val > root->val && (root->right->val < rightMax || !rightFlag)  && isValidBST(root->right, root->val, rightMax, true, rightFlag);        
        if (root->right == NULL)
            return root->left->val < root->val && (root->left->val > leftMin ||  !leftFlag) && isValidBST(root->left, leftMin, root->val, leftFlag, true);
        return root->right->val > root->val && (root->right->val < rightMax || !rightFlag) && root->left->val < root->val && (root->left->val > leftMin || !leftFlag) && isValidBST(root->left, leftMin, root->val, leftFlag, true) && isValidBST(root->right, root->val, rightMax, true, rightFlag);
    }        
};

File: Python/98. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def isValid(root, left=float('-inf'), right=float('inf')):
            if not root:
                return True
            if root.val <= left or root.val >= right:
                return False
            return isValid(root.left, left, root.val) and isValid(root.right, root.val, right)
        if not root:
            return True
        return isValid(root)
